codec-http/src/main/java/io/netty/handler/codec/http/cookie/ClientCookieEncoder.java

```diff
@@ -91,12 +121,66 @@ public String encode(Cookie... cookies) {
         }
 
         StringBuilder buf = stringBuilder();
-        for (Cookie c : cookies) {
-            if (c == null) {
-                break;
+        if (strict) {
+            ArrayList<Cookie> cookiesList = new ArrayList<Cookie>(cookies.length);
+            for (Cookie c : cookies) {
+                if (c == null) {
+                    break;
+                }
+
+                cookiesList.add(c);
             }
+            Collections.sort(cookiesList, COOKIE_COMPARATOR);
+            for (Cookie c : cookiesList) {
+                encode(buf, c);
+            }
+
+        } else {
+            for (Cookie c : cookies) {
+                if (c == null) {
+                    break;
+                }
 
-            encode(buf, c);
+                encode(buf, c);
+            }
+        }
+        return stripTrailingSeparatorOrNull(buf);
+    }
+
+    /**
+     * Encodes the specified cookies into a single Cookie header value.
+     *
+     * @param cookies some cookies
+     * @return a Rfc6265 style Cookie header value, null if no cookies are passed.
+     */
+    public String encode(Collection<? extends Cookie> cookies) {
+        if (checkNotNull(cookies, "cookies").size() == 0) {
+            return null;
+        }
+
+        StringBuilder buf = stringBuilder();
+        if (strict) {
+            ArrayList<Cookie> cookiesList = new ArrayList<Cookie>(cookies.size());
+            for (Cookie c : cookies) {
+                if (c == null) {
+                    break;
+                }
+
+                cookiesList.add(c);
+            }
+            Collections.sort(cookiesList, COOKIE_COMPARATOR);
```

Comment:
@slandelle: I didn't encounter a broken server, I'm building my own server on top of Netty, and was trying to get cookie behavior right according to the spec. I had my own cookie classes, but decided to throw them away and use raw Netty cookies, then I noticed that the Netty classes don't adhere to the spec.

The changes in the patch matter for APIs that take or return a single cookie by name, e.g. for cookie deletion (where maxAge is set to 0). This will give unexpected results if there are multiple cookies with the same name, in arbitrary order, each with different paths. You can only set (i.e. delete) one cookie at a time in the response according to the spec, and you have to pick which one to delete first if you are asked to delete a cookie based on name alone. Also it's nice to have an API that returns a single cookie with a requested name (this should be the cookie with the longest path), as opposed to returning a list of cookies, which the user has to deal with.

(I have encountered maddening problems in the past when my cookie code wasn't quite spec-compliant for other reasons, so I think it's important to get right.)

There may be other things that need fixing in the Netty code to make it spec-compliant, I don't know -- I have only been looking closely at this one issue of cookie duplication.